istioctl/cmd/upgrade-check.go

```diff
@@ -0,0 +1,100 @@
+// Copyright © 2021 NAME HERE <EMAIL ADDRESS>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/maturity"
+	"istio.io/istio/galley/pkg/config/analysis/local"
+	cfgKube "istio.io/istio/galley/pkg/config/source/kube"
+	"istio.io/istio/istioctl/pkg/clioptions"
+	"istio.io/istio/istioctl/pkg/util/formatting"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema"
+	"istio.io/istio/pkg/kube"
+)
+
+func upgradeCheckCommand() *cobra.Command {
+	var opts clioptions.ControlPlaneOptions
+	var namespaces []string
+	var allNamespaces, skipControlPlane bool
+	// cmd represents the upgradeCheck command
+	cmd := &cobra.Command{
+		Use:   "upgrade-check",
```

Comment:
do we want a new command upgrade-check, rather than just improving pre-check? I am a bit concerned about adding yet another command like this. Users will not know when to run {install, upgrade, upgrade-check, pre-check, analyze, validate, verify-install}

Reply:
pre-check is to be run before an install, and checks things like kubernetes version compatibility.  This command is focused on ensuring that your istio install is safe to upgrade.  Then again, if that's not clear to you, it's definitely not clear to our users...

Reply:
I actually *does* make sense to me, but probably not many others :slightly_smiling_face: . I think it may be best to have fewer commands. We can have a single pre-check that checks for installs and upgrades, right? If something only makes sense for "upgrade" we can disable it when there is no istio installed perhaps?

For a single concrete example, my check for binds only looks for pods with sidecars so it can work no matter what